### PR TITLE
Add resplit for subjob

### LINKED ISSUE #1669: Enable resubmission of subjobs with a new splitter
If a job is created with a splitter, it will create multiple subjobs. If some of the subjobs fail, it is possible to resubmit them, but not with a modification of the splitter. It can be very inefficient to resubmit the entire job with a new splitter. It should be possible to take a number of subjob and essentially run a new splitter on it. This might then create a number of subjobs that replace the existing one.

Imagine that we have a job where it carries out some calculation for a given integer parameter. We make a splitter that parts them such that each subjob does three calculations.

```
j = Job(splitter=ArgSplitter([[1,2,3],[4,5,6],[7,8,9]])
j.submit()
```
Now it turns out that the subjob wirh `[7,8,9]` failed because it ran out of some resource (memory, CPU time, ...) but it can be fixed by splitting it up into three separate pieces. At the moment I can only do that by creating a copy of the subjob and re-split.

```
j2 = j.subjobs[2].copy()
j2.splitter = ArgSplitter([[7],[8],[9]])
j2.submit()
```
While this works, it is unfortunate as it destroys the bookkeeping part (I now have to remember that my calculation is in both `j` and `j2`. If instead I could do

```
j.subjobs[2].splitter = ArgSplitter([[7],[8],[9]])
j.subjobs[2].submit()
```
I would end up with a single jobs object `j`, that now has 5 subjobs (I removed one by resplitting and created 3 new ones). Bookkeepoing would be much better as a result.

### PR DESCRIPTION
**Pull request by mesmith75**
Fixes #1669 , inspired by #1883 .

Looks like:
```python
[14:17:25]                                                                                                                                                                         
Ganga In [7]: newsplitter = ArgSplitter(args = newargs)                                                                                                                            

[14:17:33]                                                                                                                                                                         
Ganga In [8]: jobs(14).subjobs(0).resplit(newsplitter)                                                                                                                             
INFO     removing: /tmp/masmith/tmpejoldzk2
INFO     removing: /tmp/masmith/tmp02bfpat6
INFO     job 14 status changed to "running"
INFO     Job 14 Running PostProcessor hook
INFO     job 14 status changed to "completed"
INFO     removing: /tmp/masmith/tmpfqb7xcix
INFO     Re-splitting Job: 14.0
INFO     submitting 2 subjobs
INFO     Preparing subjobs
INFO     SubJobConfigs: 2
INFO     rjobs: 2
INFO     submitting job 14.3 to Local backend
INFO     job 14 status changed to "submitting"
INFO     job 14 status changed to "submitted"
INFO     submitting job 14.4 to Local backend
INFO     job 14 status changed to "submitting"
INFO     job 14 status changed to "submitted"
Ganga Out [8]: 1

[14:17:48]                                                                                                                                                                         
Ganga In [9]: jobs(14).subjobs                                                                                                                                                     
Ganga Out [9]: 
Registry Slice: jobs(14).subjobs (5 objects)
--------------
    fqid |    status |      name | subjobs |    application |        backend |                             backend.actualCE |                       comment |  subjob status 
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
    14.0 |completed_ |           |         |     Executable |      Localhost |                           lxplus7103.cern.ch |            - has been resplit |            --- 
    14.1 | completed |           |         |     Executable |      Localhost |                           lxplus7103.cern.ch |                               |            --- 
    14.2 | completed |           |         |     Executable |      Localhost |                           lxplus7103.cern.ch |                               |            --- 
    14.3 | completed |           |         |     Executable |      Localhost |                           lxplus7103.cern.ch |             - resplit of 14.0 |            --- 
    14.4 | completed |           |         |     Executable |      Localhost |                           lxplus7103.cern.ch |             - resplit of 14.0 |            --- 
```

### CODE REVIEW

File: ganga/GangaCore/GPIDev/Lib/Job/Job.py [hunk 1]
```diff
@@ -1457,6 +1477,111 @@ def _doSplitting(self):
...
+            self.monitorPrepare_hook(jobsubconfig)
+
+            # submit the job
+            # master_submit has been written as the interface which ganga
+            # should call, not submit directly
+
+            r = self.backend.master_submit( rjobs, jobsubconfig, jobmasterconfig)
+
+            if not r:
+                raise JobManagerError('error during submit')
```
Comment (egede) [inline ganga/GangaCore/GPIDev/Lib/Job/Job.py]: Should we remove the newly created subjobs again if there is an error in the submission?

**Review by egede — CHANGES_REQUESTED** [review]
Great to see this work. Some care will still have to be taken by a use to merge the correct output at the end, but I do not see that there is too much we can do about that.

- We should check that the initial subjob is in either a `completed` or a `failed` state. Otherwise a delayed error (and potentially an undefined state with new subjobs that are not submitted) would arise when freezing the original subjob.
- A test case should be created.

File: ganga/GangaCore/GPIDev/Lib/Job/Job.py (hunk 1 above)
Comment (alexanderrichards) [reply ganga/GangaCore/GPIDev/Lib/Job/Job.py]: probably yes, as long as we have the original they were split from

**Comment by egede** [comment]
Need to also check how this works with prepared applications.

File: ganga/GangaLHCb/Lib/RTHandlers/GaudiExecRTHandlers.py [hunk 2]
```diff
@@ -166,10 +166,10 @@ def prepareCommand(app):
                                                 getName(opts_file))
 
     #Check if this was checked out with LbEnv or not
-    isLbEnv = False
-    with open(app.directory+'/Makefile', "r") as makefile:
-        if 'LbEnv' in makefile.read():
-            isLbEnv = True
+    isLbEnv = True
```
Comment (mesmith75) [inline ganga/GangaLHCb/Lib/RTHandlers/GaudiExecRTHandlers.py]: This fixes #1897 . I am not sure this ever worked and by now LbEnv has been mandatory for so long it shouldn't be an issue to remove the check. It breaks the resubmission with a prepared application as it looks in the build directory, which may not exist anymore if the application has been submitted already.

File: ganga/GangaLHCb/Lib/RTHandlers/GaudiExecRTHandlers.py [hunk 3]
```diff
@@ -165,13 +165,7 @@ def prepareCommand(app):
...
                                                 getName(opts_file))
 
-    #Check if this was checked out with LbEnv or not
-    isLbEnv = False
-    with open(app.directory+'/Makefile', "r") as makefile:
-        if 'LbEnv' in makefile.read():
-            isLbEnv = True
-
-    sourceEnv = app.getWNEnvScript(isLbEnv)
+    sourceEnv = app.getWNEnvScript(True)
```
Comment (mesmith75) [inline ganga/GangaLHCb/Lib/RTHandlers/GaudiExecRTHandlers.py:168]: This fixes #1897 . LbEnv has been mandatory for so long now that this shouldn't be an issue.

**Review by egede — DISMISSED** [review]

**Review by egede — APPROVED** [review]
Added some tests and documentation.

File: doc/UserGuide/Splitters.rst [hunk 4]
```diff
@@ -105,3 +105,34 @@ The ``GangaDatasetSplitter`` is provided as an easy way of splitting over a numb
...
+    
+    j.subjobs(0).resplit(ArgSplitter(args=[ [1], [1] ]))
+
+    # wait for jobs to complete
+    j.subjobs
+    
+    --------------
+        fqid |    status       |                       comment |
+    -------------------------------------------------------
+         8.0 |completed_failed |            - has been resplit |
```
Comment (mesmith75) [inline doc/UserGuide/Splitters.rst:135]: Is this right? It should be `completed_frozen`. Maybe there is a typo in the code.

Comment (egede) [reply doc/UserGuide/Splitters.rst:135]: I just did a copy and paste from the screen ....

**Pull request merged**